diff -ru a/arch/um/include/asm/thread_info.h b/arch/um/include/asm/thread_info.h
--- a/arch/um/include/asm/thread_info.h
+++ b/arch/um/include/asm/thread_info.h
@@ -47,10 +47,8 @@
 /* how to get the thread information struct from C */
 static inline struct thread_info *current_thread_info(void)
 {
-	struct thread_info *ti;
-	unsigned long mask = THREAD_SIZE - 1;
-	ti = (struct thread_info *) (((unsigned long) &ti) & ~mask);
-	return ti;
+	register unsigned long sp asm("esp");
+	return (struct thread_info *)(sp & ~(THREAD_SIZE - 1));
 }
 
 #define THREAD_SIZE_ORDER CONFIG_KERNEL_STACK_ORDER
